From: Hilko Bengen <bengen@debian.org>
Date: Thu, 23 May 2013 19:20:17 +0200
Subject: perl: various fixes for separated builds

- Make sure that the .pm gets installed to the correct path
- Override Makefile.PL parameters everywhere
- Set PERL5LIB correctly for bindtests
---
 perl/Makefile.PL.in |    1 +
 perl/Makefile.am    |   24 +++++++++++++++---------
 run.in              |    4 ++--
 3 files changed, 18 insertions(+), 11 deletions(-)

diff --git a/perl/Makefile.PL.in b/perl/Makefile.PL.in
index 4f12bc0..12b56d9 100644
--- a/perl/Makefile.PL.in
+++ b/perl/Makefile.PL.in
@@ -28,4 +28,5 @@ WriteMakefile (
     INC => '-I@abs_top_builddir@/src -I@abs_top_srcdir@/src',
     TYPEMAPS => [ '@srcdir@/typemap' ],
     CCFLAGS => $Config{ccflags} . ' -DGUESTFS_PRIVATE=1 @CFLAGS@',
+    PM => {'lib/Sys/Guestfs.pm' => '@abs_builddir@/blib/arch/Sys/Guestfs.pm'},
     );
diff --git a/perl/Makefile.am b/perl/Makefile.am
index 780024f..ad5211f 100644
--- a/perl/Makefile.am
+++ b/perl/Makefile.am
@@ -63,15 +63,18 @@ TESTS_ENVIRONMENT = $(top_builddir)/run --test
 
 INSTALLDIRS = site
 
+MAKEFILE_PL_PARAMS=\
+	INST_ARCHLIB=$(abs_builddir)/blib/arch \
+	INST_SCRIPT=$(abs_builddir)/blib/script \
+	INST_BIN=$(abs_builddir)/blib/bin \
+	INST_LIB=$(abs_builddir)/blib/lib \
+	INST_MAN1DIR=$(abs_builddir)/blib/man1 \
+	INST_MAN3DIR=$(abs_builddir)/blib/man3 \
+	TEST_FILES=$(abs_srcdir)/t/*.t
+
 all: Makefile-pl src_deps
 	$(MAKE) -C $(srcdir) -f $(abs_builddir)/Makefile-pl \
-		INST_ARCHLIB=$(abs_builddir)/blib/arch \
-		INST_SCRIPT=$(abs_builddir)/blib/script \
-		INST_BIN=$(abs_builddir)/blib/bin \
-		INST_LIB=$(abs_builddir)/blib/lib \
-		INST_MAN1DIR=$(abs_builddir)/blib/man1 \
-		INST_MAN3DIR=$(abs_builddir)/blib/man3 \
-		TEST_FILES=$(abs_srcdir)/t/*.t
+		$(MAKEFILE_PL_PARAMS)
 
 Makefile-pl: Makefile.PL
 	cd $(srcdir); \
@@ -83,11 +86,14 @@ Makefile-pl: Makefile.PL
 #CLEANFILES = Makefile-pl
 
 clean-local:
-	-$(MAKE) -f Makefile-pl clean
-	rm -f Makefile-pl
+	-$(MAKE) -C $(srcdir) -f $(abs_builddir)/Makefile-pl \
+		$(MAKEFILE_PL_PARAMS) \
+		clean
+	rm -f $(abs_builddir)/Makefile-pl $(abs_builddir)/Makefile-pl.old
 
 install-data-hook:
 	$(MAKE) -C $(srcdir) -f $(abs_builddir)/Makefile-pl \
+		$(MAKEFILE_PL_PARAMS) \
 		DESTDIR=$(DESTDIR) install
 
 endif
diff --git a/run.in b/run.in
index ebdec65..9866fc7 100755
--- a/run.in
+++ b/run.in
@@ -90,9 +90,9 @@ export LD_LIBRARY_PATH
 
 # For Perl.
 if [ -z "$PERL5LIB" ]; then
-    PERL5LIB="$s/perl/lib:$b/perl/blib/lib:$b/perl/blib/arch"
+    PERL5LIB="$b/perl/lib:$b/perl/blib/lib:$b/perl/blib/arch"
 else
-    PERL5LIB="$s/perl/lib:$b/perl/blib/lib:$b/perl/blib/arch:$PERL5LIB"
+    PERL5LIB="$b/perl/lib:$b/perl/blib/lib:$b/perl/blib/arch:$PERL5LIB"
 fi
 export PERL5LIB
 
